provider_client.go

```diff
@@ -72,14 +72,29 @@ type ProviderClient struct {
 	ReauthFunc func() error
 
 	mut *sync.RWMutex
+
+	reauthmut *reauthlock
+}
+
+type reauthlock struct {
+	sync.RWMutex
+	reauthing bool
 }
 
 // AuthenticatedHeaders returns a map of HTTP headers that are common for all
 // authenticated service requests.
-func (client *ProviderClient) AuthenticatedHeaders() map[string]string {
+func (client *ProviderClient) AuthenticatedHeaders() (m map[string]string) {
+	if client.reauthmut != nil {
+		client.reauthmut.RLock()
+		if client.reauthmut.reauthing {
+			client.reauthmut.RUnlock()
```

Comment:
@jrperritt: Perhaps a slightly improved version would be:

```go
if client.reauthmut != nil {
    client.reauthmut.RLock()
    reauthing := client.reauthmut.reauthing
    client.reauthmut.RUnlock()
    if reauthing {
        return
    }
}
t := client.Token()
```

Reply:
(Removes the need for two `RUnlock` calls.)

Reply:
> Perhaps a slightly improved version would be:

Maybe? That'd work too, but it looks like the same number of lines. Note that only one of the `RUnlock`s is ever called.